src/main/java/com/minecolonies/coremod/colony/crafting/CustomRecipeFactory.java

```diff
@@ -0,0 +1,315 @@
+      final ResourceLocation lootTable, final ResourceLocation recipeId, final ResourceLocation researchReq, final ResourceLocation researchExclude, final int minBldgLevel, final int maxBldgLevel, final boolean mustExist, final boolean showTooltip)
+    {
+        return new CustomRecipe(crafter, inputs, primaryOutput, secondaryOutput, altOutputs, lootTable, recipeId, researchReq, researchExclude, minBldgLevel, maxBldgLevel, mustExist, showTooltip);
+    }
+
+    @NotNull
+    @Override
+    public TypeToken<CustomRecipe> getFactoryOutputType()
+    {
+        return TypeToken.of(CustomRecipe.class);
+    }
+
+    @NotNull
+    @Override
+    public TypeToken<FactoryVoidInput> getFactoryInputType()
+    {
+        return TypeConstants.FACTORYVOIDINPUT;
+    }
+
+    private CompoundNBT getCompoundForItemStack(final ItemStack is)
+    {
+        final CompoundNBT item = new CompoundNBT();
+        item.putString(ITEM_PROP, is.getItem().getRegistryName().toString());
+        item.putInt(COUNT_PROP, is.getCount());
+        if(is.hasTag())
+        {
+            item.put(RECIPE_TAG, is.serializeNBT());
+        }
+        return item;
+    }
+
+    private ItemStack getItemStackForCompound(final CompoundNBT nbt)
+    {
+        final ItemStack item = new ItemStack(ForgeRegistries.ITEMS.getValue(new ResourceLocation(nbt.getString(ITEM_PROP))), nbt.getInt(COUNT_PROP));
+        if(nbt.hasUniqueId(RECIPE_TAG))
+        {
+            item.deserializeNBT(nbt.getCompound(RECIPE_TAG));
+        }
+        return item;
+    }
+
+    @NotNull
+    @Override
+    public CompoundNBT serialize(@NotNull final IFactoryController controller, @NotNull final CustomRecipe recipe)
+    {
+        final CompoundNBT compound = new CompoundNBT();
+        compound.putString(RECIPE_CRAFTER_PROP, recipe.getCrafter());
+        compound.putString(CUSTOM_RECIPE_ID_PROP, recipe.getRecipeStorage().getRecipeSource().toString());
+        if(recipe.getRequiredResearchId() != null)
+        {
+            compound.putString(RECIPE_RESEARCHID_PROP, recipe.getRequiredResearchId().toString());
+        }
+        if(recipe.getExcludedResearchId() != null)
+        {
+            compound.putString(RECIPE_EXCLUDED_RESEARCHID_PROP, recipe.getExcludedResearchId().toString());
+        }
+        if(recipe.getLootTable() != null)
+        {
+            compound.putString(RECIPE_LOOTTABLE_PROP, recipe.getLootTable().toString());
+        }
+        compound.putInt(RECIPE_BUILDING_MIN_LEVEL_PROP, recipe.getMinBuildingLevel());
+        compound.putInt(RECIPE_BUILDING_MAX_LEVEL_PROP, recipe.getMaxBuildingLevel());
+        compound.putBoolean(RECIPE_MUST_EXIST, recipe.getMustExist());
+        compound.putBoolean(RECIPE_SHOW_TOOLTIP, recipe.getShowTooltip());
+        // RecipeStorage has its own factory class, but its output is larger than necessary (~1KB per recipe, vs ~800b for below, averaged over all minecolonies recipes)
```

Comment:
Can't we adjust the serialize to packetBuffer in the recipeStorage factory instead even?

Reply:
A lot of the differences between this version and the RecipeStorageFactory one are just dropped fields.  The client doesn't care about the IToken or grid size for tooltip drawing, but they may be relevant for the AddRemoveRecipeMessage (or may not be; I don't understand a lot of what RecipeStorage does on the server side).

The byte-based RecipeStorage serializer is a lot more efficient, so it doesn't really matter much; I can switch both, use the byte-based CustomRecipe serialization for the big message, and just warn that the CompoundNBT-based one will be pretty large if anyone ever uses it -- it's genuinely not that big a deal for single recipes, it's only this bulk update that could potentially burst the CompoundNBT limit. 

Using getRecipeStorage _will_ populate all of the custom recipe manager recipe storages on the server, though.  I'm not sure if it's that big a deal, but there's specific code to avoid having that cache being generated until the recipe is called for.